async process objects' stdout and stderr are not StreamReaders when using PIPE

According to the [asyncio.subprocess.PIPE docs](https://docs.python.org/3/library/asyncio-subprocess.html#asyncio.asyncio.subprocess.PIPE):
> If PIPE is passed to stdin argument, the Process.stdin attribute will point to a StreamWriter instance.
>
> If PIPE is passed to stdout or stderr arguments, the Process.stdout and Process.stderr attributes will point to StreamReader instances.

It looks like the object that fake_subprocess returns from `create_process_exec` (and probably `create_process_shell`) don't mimic this correctly. Here's an example test:
```python
import asyncio
import os

def test_stdio_and_stderr(fake_process):
    async def impl():
        process = await asyncio.create_subprocess_exec(
            "ls", "/",
            stdin=open(os.devnull, "r"), # work around #64
            stdout=asyncio.subprocess.PIPE,
            stderr=asyncio.subprocess.PIPE,
        )

        # wait for the process
        async def _read_stream(
            stream: asyncio.StreamReader, logtype: str,
        ):
            while True:
                line = await stream.readline()
                if not line:
                    break
                # do something with line

        await asyncio.wait(
            [
                asyncio.create_task(
                    _read_stream(process.stdout, "stdout")
                ),
                asyncio.create_task(
                    _read_stream(process.stderr, "stderr")
                ),
                asyncio.create_task(process.wait()),
            ]
        )

    fake_process.register_subprocess(["ls", "/"], stdout=[b"one", b"two"])
    asyncio.get_event_loop().run_until_complete(impl())

    # Force pytest to display the asyncio exception messages
    assert False
```

The expected output is for there to be a failed assertion -- but nothing else. Here's the actual output:
```
------------------------------------------- Captured log call --------------------------------------------
ERROR    asyncio:base_events.py:1707 Task exception was never retrieved
future: <Task finished name='Task-2' coro=<test_stdio_and_stderr.<locals>.impl.<locals>._read_stream() done, defined at /elm0/jranieri/src/gtirb-server2/server/test/test_stdio.py:15> exception=TypeError("object bytes can't be used in 'await' expression")>
Traceback (most recent call last):
  File "/elm0/jranieri/src/gtirb-server2/server/test/test_stdio.py", line 19, in _read_stream
    line = await stream.readline()
TypeError: object bytes can't be used in 'await' expression
ERROR    asyncio:base_events.py:1707 Task exception was never retrieved
future: <Task finished name='Task-3' coro=<test_stdio_and_stderr.<locals>.impl.<locals>._read_stream() done, defined at /elm0/jranieri/src/gtirb-server2/server/test/test_stdio.py:15> exception=TypeError("object bytes can't be used in 'await' expression")>
Traceback (most recent call last):
  File "/elm0/jranieri/src/gtirb-server2/server/test/test_stdio.py", line 19, in _read_stream
    line = await stream.readline()
TypeError: object bytes can't be used in 'await' expression
======================================== short test summary info =========================================
FAILED test_stdio.py::test_stdio_and_stderr - assert False
```

--- a/changelog.d/bug.0a9b3932.entry.yaml
+++ b/changelog.d/bug.0a9b3932.entry.yaml
@@ -1,0 +1,6 @@
+message: Make `stdout` and `stderr` an `asyncio.StreamReader` instance when using
+  asyncio functions.
+pr_ids:
+- '68'
+timestamp: 1641746962
+type: bug

--- a/pytest_subprocess/core.py
+++ b/pytest_subprocess/core.py
@@ -1,5 +1,6 @@
 # -*- coding: utf-8 -*-
 import asyncio
+import collections
 import io
 import os
 import signal
@@ -33,7 +34,7 @@ OPTIONAL_TEXT = Union[str, bytes, None]
 OPTIONAL_TEXT_OR_ITERABLE = Union[
     str, bytes, None, Sequence[Union[str, bytes]],
 ]
-BUFFER = Union[None, io.BytesIO, io.StringIO, asyncio.StreamReader]
+BUFFER = Union[io.BytesIO, io.StringIO, asyncio.StreamReader]
 ARGUMENT = Union[str, Any]
 COMMAND = Union[Sequence[ARGUMENT], str, Command]
 
@@ -45,12 +46,11 @@ class PluginInternalError(Exception):
 class FakePopen:
     """Base class that fakes the real subprocess.Popen()"""
 
-    stdout: BUFFER = None
-    stderr: BUFFER = None
+    stdout: Optional[BUFFER] = None
+    stderr: Optional[BUFFER] = None
     returncode: Optional[int] = None
     text_mode: bool = False
     pid: int = 0
-    __async__ = False
 
     def __init__(
         self,
@@ -92,12 +92,17 @@ class FakePopen:
     ) -> Tuple[AnyType, AnyType]:
         self._finalize(input, timeout)
 
+        if isinstance(self.stdout, asyncio.StreamReader) or isinstance(
+            self.stderr, asyncio.StreamReader
+        ):
+            raise PluginInternalError
+
         return (
             self.stdout.getvalue() if self.stdout else None,
             self.stderr.getvalue() if self.stderr else None,
         )
 
-    def _finalize(self, input, timeout):
+    def _finalize(self, input: OPTIONAL_TEXT, timeout: Optional[float]) -> None:
         if input and self.__stdin_callable:
             callable_output = self.__stdin_callable(input)
             if isinstance(callable_output, dict):
@@ -111,8 +116,8 @@ class FakePopen:
             self.__thread.join(timeout)
 
     def _extend_stream_from_dict(
-        self, dictionary: Dict[str, AnyType], key: str, stream: BUFFER
-    ) -> BUFFER:
+        self, dictionary: Dict[str, AnyType], key: str, stream: Optional[BUFFER]
+    ) -> Optional[BUFFER]:
         data = dictionary.get(key)
         if data:
             return self._prepare_buffer(input=data, io_base=stream)
@@ -216,18 +221,34 @@ class FakePopen:
             # both are union so could be incompatible if not _convert()
             input = io_base.getvalue() + (input)  # type: ignore
 
-        if self.__async__:
-            io_base = asyncio.StreamReader()
-            io_base.write = io_base.feed_data
-        else:
-            io_base = io.StringIO() if self.text_mode else io.BytesIO()
+        io_base = self._get_empty_buffer(self.text_mode)
 
         if input is None:
             return io_base
 
         # similar as above - mypy has to be disabled because unions
-        io_base.write(input)  # type: ignore
+        if isinstance(io_base, asyncio.StreamReader):
+            io_base.feed_data(self._data_to_bytes(input))
+        else:
+            io_base.write(input)  # type: ignore
         return io_base
+
+    def _get_empty_buffer(self, text: bool) -> BUFFER:
+        return io.StringIO() if text else io.BytesIO()
+
+    def _to_bytes(self, data: Sequence[Union[str, bytes]]) -> Sequence[bytes]:
+        return [elem if isinstance(elem, bytes) else elem.encode() for elem in data]
+
+    def _data_to_bytes(self, data: OPTIONAL_TEXT_OR_ITERABLE) -> bytes:
+        if isinstance(data, collections.Sequence) and not isinstance(data, bytes):
+            return b"\n".join(
+                (item if isinstance(item, bytes) else item.encode() for item in data)
+            )
+        if isinstance(data, str):
+            return data.encode()
+        if not data:
+            return b""
+        return data
 
     def _write_to_buffer(self, data: OPTIONAL_TEXT_OR_ITERABLE, buffer: IO) -> None:
         data_type: Callable = (
@@ -271,17 +292,14 @@ class FakePopen:
     def _finish_process(self) -> None:
         self.returncode = self.__returncode
 
-        if self.stdout:
-            if self.__async__:
-                self.stdout.feed_eof()
-            else:
-                self.stdout.seek(0)
+        self._finalize_stream(self.stdout)
+        self._finalize_stream(self.stderr)
 
-        if self.stderr:
-            if self.__async__:
-                self.stderr.feed_eof()
-            else:
-                self.stderr.seek(0)
+    def _finalize_stream(self, stream: Optional[BUFFER]) -> None:
+        if isinstance(stream, asyncio.StreamReader):
+            stream.feed_eof()
+        elif stream:
+            stream.seek(0)
 
     def received_signals(self) -> Tuple[int, ...]:
         """Get a tuple of signals received by the process."""
@@ -291,7 +309,8 @@ class FakePopen:
 class AsyncFakePopen(FakePopen):
     """Class to handle async processes"""
 
-    __async__ = True
+    stdout: asyncio.StreamReader
+    stderr: asyncio.StreamReader
 
     async def communicate(  # type: ignore
         self, input: OPTIONAL_TEXT = None, timeout: Optional[float] = None
@@ -306,6 +325,9 @@ class AsyncFakePopen(FakePopen):
 
     async def wait(self, timeout: Optional[float] = None) -> int:  # type: ignore
         return super().wait(timeout)
+
+    def _get_empty_buffer(self, _: bool) -> asyncio.StreamReader:
+        return asyncio.StreamReader()
 
 
 class ProcessNotRegisteredError(Exception):

--- a/pytest_subprocess/core.py
+++ b/pytest_subprocess/core.py
@@ -1,6 +1,6 @@
 # -*- coding: utf-8 -*-
 import asyncio
-import collections
+import collections.abc
 import io
 import os
 import signal
@@ -240,7 +240,7 @@ class FakePopen:
         return [elem if isinstance(elem, bytes) else elem.encode() for elem in data]
 
     def _data_to_bytes(self, data: OPTIONAL_TEXT_OR_ITERABLE) -> bytes:
-        if isinstance(data, collections.Sequence) and not isinstance(data, bytes):
+        if isinstance(data, collections.abc.Sequence) and not isinstance(data, bytes):
             return b"\n".join(
                 (item if isinstance(item, bytes) else item.encode() for item in data)
             )

--- a/pytest_subprocess/core.py
+++ b/pytest_subprocess/core.py
@@ -90,7 +90,8 @@ class FakePopen:
     def communicate(
         self, input: OPTIONAL_TEXT = None, timeout: Optional[float] = None
     ) -> Tuple[AnyType, AnyType]:
-        self._finalize(input, timeout)
+        self._handle_stdin(input)
+        self._finalize_thread(timeout)
 
         if isinstance(self.stdout, asyncio.StreamReader) or isinstance(
             self.stderr, asyncio.StreamReader
@@ -102,7 +103,7 @@ class FakePopen:
             self.stderr.getvalue() if self.stderr else None,
         )
 
-    def _finalize(self, input: OPTIONAL_TEXT, timeout: Optional[float]) -> None:
+    def _handle_stdin(self, input: OPTIONAL_TEXT) -> None:
         if input and self.__stdin_callable:
             callable_output = self.__stdin_callable(input)
             if isinstance(callable_output, dict):
@@ -112,6 +113,8 @@ class FakePopen:
                 self.stderr = self._extend_stream_from_dict(
                     callable_output, "stderr", self.stderr
                 )
+
+    def _finalize_thread(self, timeout: Optional[float]) -> None:
         if self.__thread is not None:
             self.__thread.join(timeout)
 
@@ -216,12 +219,17 @@ class FakePopen:
         if input and self.__universal_newlines and isinstance(input, str):
             input = input.replace("\r\n", "\n")
 
-        if io_base is not None:
+        if (
+            io_base
+            and not isinstance(io_base, asyncio.StreamReader)
+            and io_base.tell() == 0
+        ):
             # same reason for disabling mypy as in `input = linesep.join...`:
             # both are union so could be incompatible if not _convert()
             input = io_base.getvalue() + (input)  # type: ignore
 
-        io_base = self._get_empty_buffer(self.text_mode)
+        if io_base is None:
+            io_base = self._get_empty_buffer(self.text_mode)
 
         if input is None:
             return io_base
@@ -316,7 +324,8 @@ class AsyncFakePopen(FakePopen):
         self, input: OPTIONAL_TEXT = None, timeout: Optional[float] = None
     ) -> Tuple[AnyType, AnyType]:
 
-        self._finalize(input, timeout)
+        self._handle_stdin(input)
+        self._finalize_thread(timeout)
 
         return (
             await self.stdout.read() if self.stdout else None,

--- a/pytest_subprocess/core.py
+++ b/pytest_subprocess/core.py
@@ -90,6 +90,14 @@ class FakePopen:
     def communicate(
         self, input: OPTIONAL_TEXT = None, timeout: Optional[float] = None
     ) -> Tuple[AnyType, AnyType]:
+        self._finalize(input, timeout)
+
+        return (
+            self.stdout.getvalue() if self.stdout else None,
+            self.stderr.getvalue() if self.stderr else None,
+        )
+
+    def _finalize(self, input, timeout):
         if input and self.__stdin_callable:
             callable_output = self.__stdin_callable(input)
             if isinstance(callable_output, dict):
@@ -99,14 +107,8 @@ class FakePopen:
                 self.stderr = self._extend_stream_from_dict(
                     callable_output, "stderr", self.stderr
                 )
-
         if self.__thread is not None:
             self.__thread.join(timeout)
-
-        return (
-            self.stdout.getvalue() if self.stdout else None,
-            self.stderr.getvalue() if self.stderr else None,
-        )
 
     def _extend_stream_from_dict(
         self, dictionary: Dict[str, AnyType], key: str, stream: BUFFER
@@ -277,9 +279,9 @@ class FakePopen:
 
         if self.stderr:
             if self.__async__:
-                self.stderr.seek(0)
-            else:
                 self.stderr.feed_eof()
+            else:
+                self.stderr.seek(0)
 
     def received_signals(self) -> Tuple[int, ...]:
         """Get a tuple of signals received by the process."""
@@ -294,7 +296,13 @@ class AsyncFakePopen(FakePopen):
     async def communicate(  # type: ignore
         self, input: OPTIONAL_TEXT = None, timeout: Optional[float] = None
     ) -> Tuple[AnyType, AnyType]:
-        return super().communicate(input, timeout)
+
+        self._finalize(input, timeout)
+
+        return (
+            await self.stdout.read() if self.stdout else None,
+            await self.stderr.read() if self.stderr else None,
+        )
 
     async def wait(self, timeout: Optional[float] = None) -> int:  # type: ignore
         return super().wait(timeout)

--- a/pytest_subprocess/core.py
+++ b/pytest_subprocess/core.py
@@ -33,7 +33,7 @@ OPTIONAL_TEXT = Union[str, bytes, None]
 OPTIONAL_TEXT_OR_ITERABLE = Union[
     str, bytes, None, Sequence[Union[str, bytes]],
 ]
-BUFFER = Union[None, io.BytesIO, io.StringIO]
+BUFFER = Union[None, io.BytesIO, io.StringIO, asyncio.StreamReader]
 ARGUMENT = Union[str, Any]
 COMMAND = Union[Sequence[ARGUMENT], str, Command]
 
@@ -50,6 +50,7 @@ class FakePopen:
     returncode: Optional[int] = None
     text_mode: bool = False
     pid: int = 0
+    __async__ = False
 
     def __init__(
         self,
@@ -185,7 +186,7 @@ class FakePopen:
 
     def _prepare_buffer(
         self, input: OPTIONAL_TEXT_OR_ITERABLE, io_base: BUFFER = None,
-    ) -> Union[io.BytesIO, io.StringIO]:
+    ) -> Union[io.BytesIO, io.StringIO, asyncio.StreamReader]:
         linesep = self._convert(os.linesep)
 
         if isinstance(input, (list, tuple)):
@@ -213,7 +214,11 @@ class FakePopen:
             # both are union so could be incompatible if not _convert()
             input = io_base.getvalue() + (input)  # type: ignore
 
-        io_base = io.StringIO() if self.text_mode else io.BytesIO()
+        if self.__async__:
+            io_base = asyncio.StreamReader()
+            io_base.write = io_base.feed_data
+        else:
+            io_base = io.StringIO() if self.text_mode else io.BytesIO()
 
         if input is None:
             return io_base
@@ -265,10 +270,16 @@ class FakePopen:
         self.returncode = self.__returncode
 
         if self.stdout:
-            self.stdout.seek(0)
+            if self.__async__:
+                self.stdout.feed_eof()
+            else:
+                self.stdout.seek(0)
 
         if self.stderr:
-            self.stderr.seek(0)
+            if self.__async__:
+                self.stderr.seek(0)
+            else:
+                self.stderr.feed_eof()
 
     def received_signals(self) -> Tuple[int, ...]:
         """Get a tuple of signals received by the process."""
@@ -277,6 +288,8 @@ class FakePopen:
 
 class AsyncFakePopen(FakePopen):
     """Class to handle async processes"""
+
+    __async__ = True
 
     async def communicate(  # type: ignore
         self, input: OPTIONAL_TEXT = None, timeout: Optional[float] = None

--- a/tests/test_asyncio.py
+++ b/tests/test_asyncio.py
@@ -201,22 +201,31 @@ async def test_stdio_and_stderr(fake_process, fake):
 
     if fake:
         fake_process.register_subprocess(
-            ["python", "example_script.py"], stdout="Stdout line 1\nStdout line 2",
+            ["python", "example_script.py", "stderr"],
+            stdout="Stdout line 1\nStdout line 2",
+            stderr="Stderr line 1",
         )
     else:
         fake_process.allow_unregistered(True)
 
     process = await asyncio.create_subprocess_exec(
-        "python", "example_script.py", stdout=asyncio.subprocess.PIPE
+        "python",
+        "example_script.py",
+        "stderr",
+        stdout=asyncio.subprocess.PIPE,
+        stderr=asyncio.subprocess.PIPE,
     )
 
     stdout_list = []
+    stderr_list = []
     await asyncio.gather(
         asyncio.create_task(_read_stream(process.stdout, stdout_list)),
+        asyncio.create_task(_read_stream(process.stderr, stderr_list)),
         asyncio.create_task(process.wait()),
     )
 
     assert len(stdout_list) == 2
+    assert len(stderr_list) == 1
 
 
 @pytest.fixture(autouse=True)

--- a/tests/test_asyncio.py
+++ b/tests/test_asyncio.py
@@ -188,6 +188,37 @@ async def test_anyio(fake_process):
     await anyio.sleep(0.01)
 
 
+@pytest.mark.asyncio
+@pytest.mark.parametrize("fake", [False, True])
+async def test_stdio_and_stderr(fake_process, fake):
+    async def _read_stream(stream: asyncio.StreamReader, output_list):
+        while True:
+            line = await stream.readline()
+            if not line:
+                break
+            else:
+                output_list.append(line)
+
+    if fake:
+        fake_process.register_subprocess(
+            ["python", "example_script.py"], stdout="Stdout line 1\nStdout line 2",
+        )
+    else:
+        fake_process.allow_unregistered(True)
+
+    process = await asyncio.create_subprocess_exec(
+        "python", "example_script.py", stdout=asyncio.subprocess.PIPE
+    )
+
+    stdout_list = []
+    await asyncio.gather(
+        asyncio.create_task(_read_stream(process.stdout, stdout_list)),
+        asyncio.create_task(process.wait()),
+    )
+
+    assert len(stdout_list) == 2
+
+
 @pytest.fixture(autouse=True)
 def skip_on_pypy():
     """Async test for some reason crash on pypy 3.6 on Windows"""

--- a/tests/test_asyncio.py
+++ b/tests/test_asyncio.py
@@ -252,10 +252,12 @@ async def test_combined_stdout_and_stderr(fake_process, fake):
         loop.create_task(process.wait()),
     )
 
-    assert stdout_list == [
+    # sorted() is necessary here, as the order here may be not deterministic,
+    # and sometimes stderr comes before stdout or the opposite
+    assert sorted(stdout_list) == [
+        f"Stderr line 1{os.linesep}",
         f"Stdout line 1{os.linesep}",
         f"Stdout line 2{os.linesep}",
-        f"Stderr line 1{os.linesep}",
     ]
     assert stderr_list == []
 

--- a/tests/test_asyncio.py
+++ b/tests/test_asyncio.py
@@ -259,11 +259,12 @@ async def test_combined_stdout_and_stderr(fake_process, fake):
         f"Stdout line 1{os.linesep}",
         f"Stdout line 2{os.linesep}",
     ]
-    assert stderr_list == []
+    assert stderr_list == ["empty"]
 
 
 async def _read_stream(stream: asyncio.StreamReader, output_list):
     if stream is None:
+        output_list.append("empty")
         return None
 
     while True:

--- a/tests/test_asyncio.py
+++ b/tests/test_asyncio.py
@@ -218,10 +218,12 @@ async def test_stdio_and_stderr(fake_process, fake):
 
     stdout_list = []
     stderr_list = []
+
+    loop = asyncio.get_event_loop()
     await asyncio.gather(
-        asyncio.create_task(_read_stream(process.stdout, stdout_list)),
-        asyncio.create_task(_read_stream(process.stderr, stderr_list)),
-        asyncio.create_task(process.wait()),
+        loop.create_task(_read_stream(process.stdout, stdout_list)),
+        loop.create_task(_read_stream(process.stderr, stderr_list)),
+        loop.create_task(process.wait()),
     )
 
     assert len(stdout_list) == 2

--- a/tests/test_asyncio.py
+++ b/tests/test_asyncio.py
@@ -190,20 +190,12 @@ async def test_anyio(fake_process):
 
 @pytest.mark.asyncio
 @pytest.mark.parametrize("fake", [False, True])
-async def test_stdio_and_stderr(fake_process, fake):
-    async def _read_stream(stream: asyncio.StreamReader, output_list):
-        while True:
-            line = await stream.readline()
-            if not line:
-                break
-            else:
-                output_list.append(line)
-
+async def test_stdout_and_stderr(fake_process, fake):
     if fake:
         fake_process.register_subprocess(
             ["python", "example_script.py", "stderr"],
-            stdout="Stdout line 1\nStdout line 2",
-            stderr="Stderr line 1",
+            stdout=["Stdout line 1", "Stdout line 2"],
+            stderr=["Stderr line 1"],
         )
     else:
         fake_process.allow_unregistered(True)
@@ -226,8 +218,58 @@ async def test_stdio_and_stderr(fake_process, fake):
         loop.create_task(process.wait()),
     )
 
-    assert len(stdout_list) == 2
-    assert len(stderr_list) == 1
+    assert stdout_list == [f"Stdout line 1{os.linesep}", f"Stdout line 2{os.linesep}"]
+    assert stderr_list == [f"Stderr line 1{os.linesep}"]
+
+
+@pytest.mark.asyncio
+@pytest.mark.parametrize("fake", [False, True])
+async def test_combined_stdout_and_stderr(fake_process, fake):
+    if fake:
+        fake_process.register_subprocess(
+            ["python", "example_script.py", "stderr"],
+            stdout=["Stdout line 1", "Stdout line 2"],
+            stderr=["Stderr line 1"],
+        )
+    else:
+        fake_process.allow_unregistered(True)
+
+    process = await asyncio.create_subprocess_exec(
+        "python",
+        "example_script.py",
+        "stderr",
+        stdout=asyncio.subprocess.PIPE,
+        stderr=asyncio.subprocess.STDOUT,
+    )
+
+    stdout_list = []
+    stderr_list = []
+
+    loop = asyncio.get_event_loop()
+    await asyncio.gather(
+        loop.create_task(_read_stream(process.stdout, stdout_list)),
+        loop.create_task(_read_stream(process.stderr, stderr_list)),
+        loop.create_task(process.wait()),
+    )
+
+    assert stdout_list == [
+        f"Stdout line 1{os.linesep}",
+        f"Stdout line 2{os.linesep}",
+        f"Stderr line 1{os.linesep}",
+    ]
+    assert stderr_list == []
+
+
+async def _read_stream(stream: asyncio.StreamReader, output_list):
+    if stream is None:
+        return None
+
+    while True:
+        line = await stream.readline()
+        if not line:
+            break
+        else:
+            output_list.append(line.decode())
 
 
 @pytest.fixture(autouse=True)

